File: backend/brackets/bracket_probability.py
from __future__ import annotations

import math
from typing import Dict

from .bracket_decode import bracket_from_int
from .elo import elo_win_probability


def bracket_log_probability(n: int, elo_ratings: Dict[str, float]) -> float:
    """
    Natural log of bracket probability (product over 63 games).

    This avoids massive underflow that would make probabilities become 0.0
    in float arithmetic.
    """

    bracket = bracket_from_int(n)
    bits = bracket["bits"]
    games = bracket["games"]

    # Prevent log(0) due to extreme floating point underflow.
    eps = 1e-300
    log_prob = 0.0

    for game_idx, game in enumerate(games):
        top_slug = game["top_team"]["slug"]
        bottom_slug = game["bottom_team"]["slug"]
        top_elo = float(elo_ratings[top_slug])
        bottom_elo = float(elo_ratings[bottom_slug])

        p_top = elo_win_probability(top_elo, bottom_elo)
        # Numerical safety: clamp to (eps, 1-eps)
        p_top = min(max(p_top, eps), 1.0 - eps)

        # bit=0 => top wins, bit=1 => bottom wins
        bit = bits[game_idx]
        prob_game = p_top if bit == 0 else (1.0 - p_top)
        prob_game = min(max(prob_game, eps), 1.0 - eps)
        log_prob += math.log(prob_game)

    return log_prob


def bracket_probability(n: int, elo_ratings: Dict[str, float]) -> float:
    """Bracket probability as a float (may underflow to 0.0)."""

    log_prob = bracket_log_probability(n, elo_ratings)
    return math.exp(log_prob)

File: frontend/src/engine/bracketProbability.js
// bracketProbability.js
//
// Mirrors `backend/brackets/bracket_engine.py`:
// - bracket_log_probability(n, elo_ratings)
// - bracket_probability(n, elo_ratings)

import { eloWinProbability } from "./elo.js";
import { bracketFromInt } from "./bracketDecode.js";
import { getSelectedTournament } from "./tournamentState.js";

export function bracketLogProbability(n, eloRatings) {
  // Product over all 63 games, represented as sum of logs to avoid underflow.
  const bracket = bracketFromInt(n, getSelectedTournament());
  const bits = bracket.bits;
  let logProb = 0.0;

  const eps = 1e-300; // numerical safety

  for (let gameIdx = 0; gameIdx < bracket.games.length; gameIdx++) {
    const game = bracket.games[gameIdx];
    const topSlug = game.top_team.slug;
    const bottomSlug = game.bottom_team.slug;

    const topElo = eloRatings[topSlug];
    const bottomElo = eloRatings[bottomSlug];
    if (topElo === undefined || bottomElo === undefined) {
      throw new Error("Missing elo rating for one of the teams");
    }

    const pTopRaw = eloWinProbability(topElo, bottomElo);
    const pTop = Math.min(Math.max(pTopRaw, eps), 1.0 - eps);

    const bit = bits[gameIdx];
    const probGame = bit === 0 ? pTop : 1.0 - pTop;
    const probClamped = Math.min(Math.max(probGame, eps), 1.0 - eps);
    logProb += Math.log(probClamped);
  }

  return logProb;
}

export function bracketProbability(n, eloRatings) {
  // WARNING: may underflow to 0 for unlikely brackets.
  const logP = bracketLogProbability(n, eloRatings);
  return Math.exp(logP);
}

